### ISSUE TITLE
Various DB points in regards to improving the indexing stage (not a bug)

### ISSUE THREAD

**Issue by megalepozy** [blockscout/blockscout#9236]
### Description

Hi,

First of all sorry for reporting this as a bug as this is not a bug, I'm just not sure how/where to report this.

1. There are 2 indexes which are almost certainly not useful:
    -  "transactions_block_consensus_index" btree (block_consensus) - is a full index on a boolean value, which result in every value indexed while the index will be used only if one of the possible 3 values will have a "small" representation of the total. It is almost certainly better to change it to a partial index if not dropping it altogether. Proper explanation can be read at https://stackoverflow.com/questions/42972726/postgresql-create-index-for-boolean-column 
    - "tokens_contract_address_hash_index" UNIQUE, btree (contract_address_hash) - which on its own is OK but the primary key index from the same table is also on the contract_address_hash column which is also set to not null, so essentially there is no difference between the 2.
2. A big number of indexes are not used at all or almost never at the indexing (indexer) stage (or at least weren't used yet at my system which I run for few days now and indexed ~600k blocks from ETH mainnet).
    - Maybe they will be used at a later indexing stage? (maybe when blocks with consensus false will need to be refetched?)
    - I guess that most of them are made for serving the frontend queries
    - The problem with those indexes is that they slow the indexing process considerably, and actually take more space than the tables data themselves at many cases... if they r needed, maybe a flag can be defined to create them only after the indexer finished the initial indexing process?
    - Currently I removed the following 52 indexes from my system and it runs way faster (still slow.... working on it and will open another issue about it ;)):
```
DROP INDEX logs_first_topic_index;
DROP INDEX logs_second_topic_index ;
DROP INDEX logs_third_topic_index ;
DROP INDEX logs_fourth_topic_index ;
DROP INDEX logs_address_hash_index;
DROP INDEX logs_transaction_hash_index_index;
DROP INDEX logs_address_hash_transaction_hash_index;
DROP INDEX "logs_block_number_DESC__index_DESC_index";
DROP INDEX "logs_block_number_ASC__index_ASC_index";
DROP INDEX logs_index_index;

DROP INDEX address_coin_balances_block_number_index;
DROP INDEX address_coin_balances_value_fetched_at_index;

DROP INDEX internal_transactions_from_address_hash_index; 
DROP INDEX internal_transactions_transaction_hash_index_index; 
DROP INDEX internal_transactions_from_address_hash_partial_index; 
DROP INDEX internal_transactions_to_address_hash_partial_index; 
DROP INDEX internal_transactions_created_contract_address_hash_partial_ind; 

DROP INDEX address_token_balances_block_number_address_hash_index;
DROP INDEX address_token_balances_token_id_index;
DROP INDEX address_token_balances_address_hash_token_contract_address_hash;
DROP INDEX address_token_balances_token_contract_address_hash_index;
-- would love to drop address_token_balances_pkey as well as it is not used, but its a primary key so I avoid it for now

DROP INDEX token_transfers_to_address_hash_transaction_hash_index;
DROP INDEX token_transfers_from_address_hash_transaction_hash_index;
DROP INDEX token_transfers_token_contract_address_hash_transaction_hash_in;
DROP INDEX "token_transfers_block_number_DESC_log_index_DESC_index";
DROP INDEX token_transfers_token_contract_address_hash_block_number_index;
DROP INDEX "token_transfers_block_number_ASC_log_index_ASC_index";
DROP INDEX token_transfers_token_contract_address_hash_token_ids_index;
DROP INDEX token_transfers_block_number_index;
DROP INDEX token_transfers_transaction_hash_log_index_index;

DROP INDEX transactions_inserted_at_index;
DROP INDEX transactions_updated_at_index;
DROP INDEX transactions_block_hash_index_index;
DROP INDEX method_id;
DROP INDEX transactions_from_address_hash_with_pending_index;
DROP INDEX "transactions_to_address_hash_with_pending_index_asc";
DROP INDEX transactions_to_address_hash_with_pending_index;
DROP INDEX transactions_block_timestamp_index;
DROP INDEX transactions_block_consensus_index;
DROP INDEX transactions_recent_collated_index;
DROP INDEX transactions_created_contract_code_indexed_at_index;

DROP INDEX address_current_token_balances_token_contract_address_hash_inde;
DROP INDEX address_current_token_balances_token_contract_address_hash_valu;
DROP INDEX address_current_token_balances_token_id_index;
DROP INDEX address_cur_token_balances_index;
-- address_current_token_balances_pkey is also a primary key index which is not used so I don't drop it yet

DROP INDEX addresses_fetched_coin_balance_hash_index;
DROP INDEX addresses_fetched_coin_balance_index;

// ERROR:  cannot drop index tokens_contract_address_hash_index because other objects depend on it
// DETAIL:  constraint token_instances_token_contract_address_hash_fkey on table token_instances depends on index tokens_contract_address_hash_index
BEGIN;
ALTER TABLE token_instances DROP CONSTRAINT token_instances_token_contract_address_hash_fkey;
DROP INDEX tokens_contract_address_hash_index; 
ALTER TABLE token_instances ADD CONSTRAINT token_instances_token_contract_address_hash_fkey FOREIGN KEY (token_contract_address_hash) REFERENCES tokens(contract_address_hash);
COMMIT;

DROP INDEX tokens_symbol_index;
DROP INDEX tokens_trgm_idx;

DROP INDEX token_instances_token_id_index;
DROP INDEX token_instances_owner_address_hash_index;

-- btw I checked if the indexes are used or not by running:
SELECT relname , indexrelname , idx_scan , idx_tup_read , idx_tup_fetch FROM pg_stat_user_indexes WHERE schemaname = 'public' and relname IN ('transactions', 'addresses', 'address_coin_balances', 'address_current_token_balances', 'address_token_balances', 'internal_transactions', 'logs', 'token_transfers', 'tokens');

-- Another nice query to see the size of the various indexes:
SELECT i.relname "Table Name",indexrelname "Index Name",
 pg_size_pretty(pg_total_relation_size(relid)) As "Total Size",
 pg_size_pretty(pg_indexes_size(relid)) as "Total Size of all Indexes",
 pg_size_pretty(pg_relation_size(relid)) as "Table Size",
 pg_size_pretty(pg_relation_size(indexrelid)) "Index Size",
 reltuples::bigint "Estimated table row count"
 FROM pg_stat_all_indexes i JOIN pg_class c ON i.relid=c.oid 
 WHERE i.relname='logs'; 
```
3. Another flag which may be useful (at least for my needs ;)) can indicate which indexes are needed (solo db population, API, full (inc. frontend), etc...) and create only the needed ones 
4. Can the DB be updated to version 16?

Sorry for cramming so many points into one issue... asked at Discord and this was what I was recommended to do, let me know if you prefer it in another way.

Thanks for this amazing project!

### Type of the installation

Docker-compose

### Type of the JSON RPC archive node

Erigon

### Type of the chain

Ethereum mainnet

### Backend version

v6.0.0

### Frontend version

not relevant

### Elixir & Erlang/OTP versions

whatever was installed with Erigon

### Operating system

Linux

**Comment by vbaranov** [comment]
We continuously monitor and optimize unused indexes. The latest version (v9 engine) includes significant improvements in terms of optimized indices set. Please update your setup to this version.
